### PR TITLE
[Stream]  sync protocol

### PR DESCRIPTION
**Pull request by JackyWYX**
## Description

This is a break down pr of https://github.com/harmony-one/harmony/pull/3535 and is rebased on https://github.com/harmony-one/harmony/pull/3582.

* Added sync stream handling code logic.
* Both client and server side logic is in function `syncStream.run`
* Client side logic is called from upper function calls (downloader module) through methods defined in `client.go`

### Commit signature and bitmaps

Currently, block commit sig and bitmap written in db is defined as the concatenation of `signature` + `bitmap`. But blocks long time ago seems having a different data written in db. So currently the way work around to get the correct commit sig and bitmap is through:

1. Try to get the next header.
2. If the next header exist, get signature and bitmaps from the next header and write this data to client.
3. Else read the signature and bitmap from db and write to client.

Please let me know if there is any problem with this approach. Thanks.

### CODE REVIEW

**Comment by JackyWYX** [comment]
Following is the data I collected for block number 1 during testing:

https://gist.github.com/JackyWYX/adba7bf16bccce479045600660d8454b

Had no idea how to interpret the commit sig written in DB.

**Comment by rlan35** [comment]
Always use the sig data from next block's header if the current block is not the latest block.

**Comment by JackyWYX** [comment]
> Always use the sig data from next block's header if the current block is not the latest block.

Cool! Yea, this is exactly the logic implemented. Please kindly review. Thanks.

File: p2p/stream/protocols/sync/chain.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+	targetBN := ch.schedule.EpochLastBlock(epoch - 1)
+	res.Header = ch.chain.GetHeaderByNumber(targetBN)
+	if res.Header == nil {
+		// we still don't have the given epoch
+		return res, nil
+	}
+	epochBI := new(big.Int).SetUint64(epoch)
+	if ch.chain.Config().IsPreStaking(epochBI) {
+		// For epoch before preStaking, only hash is stored in header
```
Comment (rlan35) [inline p2p/stream/protocols/sync/chain.go:158]: what does this mean? only hash is stored in header?

File: p2p/stream/protocols/sync/client.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+	blocks, err := req.getBlocksFromResponse(resp)
+	if err != nil {
+		return nil, stid, err
+	}
+	return blocks, stid, nil
+}
+
+// GetEpochState get the epoch block from querying the remote node running sync stream protocol.
+// Currently, this method is only supported by beacon syncer.
+// Note: use this after epoch chain is implemented.
```
Comment (rlan35) [inline p2p/stream/protocols/sync/client.go:43]: This is only used for sync'ing epoch chain right? it shouldn't be needed more normal blocks sync

File: p2p/stream/protocols/sync/chain.go (hunk 1 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/chain.go:158]: My bad. This was experimental code that wasn't verified. Let me remove these.

File: p2p/stream/protocols/sync/client.go (hunk 2 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:43]: Right. Let me remove this.

**Review by rlan35 — APPROVED** [review]

File: p2p/stream/protocols/sync/client.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+	}
+	blocks := make([]*types.Block, 0, len(blockBytes))
+	for i, bb := range blockBytes {
+		var block *types.Block
+		if err := rlp.DecodeBytes(bb, &block); err != nil {
+			return nil, errors.Wrap(err, "[GetBlocksByHashesResponse]")
+		}
+		if block != nil {
+			block.SetCurrentCommitSig(sigs[i])
+		}
```
Comment (LeoHChen) [inline p2p/stream/protocols/sync/client.go:340]: what happens if block is nil?  seems you keep adding it?

File: p2p/stream/protocols/sync/client.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+	blocks := make([]*types.Block, 0, len(blockBytes))
+	for i, bb := range blockBytes {
+		var block *types.Block
+		if err := rlp.DecodeBytes(bb, &block); err != nil {
+			return nil, errors.Wrap(err, "[GetBlocksByNumResponse]")
+		}
+		if block != nil {
+			block.SetCurrentCommitSig(sigs[i])
+		}
+		blocks = append(blocks, block)
```
Comment (LeoHChen) [inline p2p/stream/protocols/sync/client.go:152]: if block is nil, keep adding it?

File: p2p/stream/protocols/sync/const.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package sync
+
+import "time"
+
+const (
+	// GetBlockHashesAmountCap is the cap of GetBlockHashes reqeust
+	GetBlockHashesAmountCap = 50
+
+	// GetBlocksByNumAmountCap is the cap of request of a single GetBlocksByNum request
+	GetBlocksByNumAmountCap = 10
```
Comment (LeoHChen) [inline p2p/stream/protocols/sync/const.go:10]: can you add some additional comments here why 10?  I remember there was a max size limit.

File: p2p/stream/protocols/sync/protocol.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	smConfig := streammanager.Config{
+		SoftLoCap: config.SmSoftLowCap,
+		HardLoCap: config.SmHardLowCap,
+		HiCap:     config.SmHiCap,
+		DiscBatch: config.DiscBatch,
+	}
+	sp.sm = streammanager.NewStreamManager(sp.ProtoID(), config.Host, config.Discovery,
+		sp.HandleStream, smConfig)
+
+	sp.rl = ratelimiter.NewRateLimiter(sp.sm, 50, 10)
```
Comment (LeoHChen) [inline p2p/stream/protocols/sync/protocol.go:94]: comment on the 50, 10? or make them constants?

File: p2p/stream/protocols/sync/protocol.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+		if e != nil {
+			p.logger.Warn().Err(e).Str("protocol", string(pid)).
+				Msg("cannot advertise sync protocol")
+			continue
+		}
+		if nextWait == 0 || nextWait > w {
+			nextWait = w
+		}
+	}
+	if nextWait < minAdvertiseInterval {
```
Comment (LeoHChen) [inline p2p/stream/protocols/sync/protocol.go:196]: advertise every minute?

File: p2p/stream/protocols/sync/stream.go [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+		if err != nil {
+			return nil, err
+		}
+		blocksBytes = append(blocksBytes, bb)
+
+		var sig []byte
+		if block != nil {
+			sig = block.GetCurrentCommitSig()
+		}
+		sigs = append(sigs, sig)
```
Comment (LeoHChen) [inline p2p/stream/protocols/sync/stream.go:305]: what if block is nil?

**Review by LeoHChen — COMMENTED** [review]
Sorry for the late review. LGTM in general, just a few points need clarification.  Please also add metrics later when handling each message.

File: p2p/stream/protocols/sync/chain.go (hunk 1 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/chain.go]: Code removed

File: p2p/stream/protocols/sync/client.go (hunk 2 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go]: Code removed

File: p2p/stream/protocols/sync/client.go [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+)
+
+// GetBlocksByNumber do getBlocksByNumberRequest through sync stream protocol.
+// Return the block as result, target stream id, and error
+func (p *Protocol) GetBlocksByNumber(ctx context.Context, bns []uint64, opts ...Option) ([]*types.Block, sttypes.StreamID, error) {
+	if len(bns) > GetBlocksByNumAmountCap {
+		return nil, "", fmt.Errorf("number of blocks exceed cap of %v", GetBlocksByNumAmountCap)
+	}
+
+	req := newGetBlocksByNumberRequest(bns)
```
[the remark this answers is not in the document]
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:25]: You are right. Better do a sanity check and not send out the nil request if empty. Logics added.

File: p2p/stream/protocols/sync/client.go [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+	resp, stid, err := p.rm.DoRequest(ctx, req, opts...)
+	if err != nil {
+		// At this point, error can be context canceled, context timed out, or waiting queue
+		// is already full.
+		return nil, stid, err
+	}
+
+	// Parse and return blocks
+	blocks, err := req.getBlocksFromResponse(resp)
```
[the remark this answers is not in the document]
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:34]: The function is blocked at 

```
resp, stid, err := p.rm.DoRequest(ctx, req, opts...)
```

And this step is purely extracting the result from the response. Whether it is a sync response is also check within this function. Please take a look.

File: p2p/stream/protocols/sync/client.go [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+	}
+	blockBytes, sigs, err := req.parseBlockBytesAndSigs(sResp)
+	if err != nil {
+		return nil, err
+	}
+	blocks := make([]*types.Block, 0, len(blockBytes))
+	for i, bb := range blockBytes {
+		var block *types.Block
+		if err := rlp.DecodeBytes(bb, &block); err != nil {
+			return nil, errors.Wrap(err, "[GetBlocksByNumResponse]")
```
[the remark this answers is not in the document]
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:147]: 1. Why the error message is not clear? What is your expectation of the error message? (Currently, the error message is like `[GetBlocksByNumResponse] rlp: some rlp error message)
2. Yes and no. All blocks need to be correctly RLP decoded here. If decode errored here, I won't return any blocks but the criminal stream ID accused of delivering this false message. But the overall validation of the block message and error handling logic is executed in downloader module.

File: p2p/stream/protocols/sync/client.go [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+	sResp, ok := resp.(*syncResponse)
+	if !ok || sResp == nil {
+		return nil, errors.New("not sync response")
+	}
+	if errResp := sResp.pb.GetErrorResponse(); errResp != nil {
+		return nil, errors.New(errResp.Error)
+	}
+	bhResp := sResp.pb.GetGetBlocksByHashesResponse()
+	if bhResp == nil {
+		return nil, errors.New("response not GetBlocksByHashes")
```
[the remark this answers is not in the document]
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:323]: What is your expectation of this error message? 

For example, the remote node delivers GetBlocksByNumber response for GetBlocksByHashes request. So this error reports the response type is not correct. Actually, this error message shall never be triggered if the remote node is honest.

File: p2p/stream/protocols/sync/client.go [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+		sigs       = bhResp.CommitSig
+	)
+	if len(blockBytes) != len(sigs) {
+		return nil, fmt.Errorf("sig size not expected: %v / %v", len(sigs), len(blockBytes))
+	}
+	blocks := make([]*types.Block, 0, len(blockBytes))
+	for i, bb := range blockBytes {
+		var block *types.Block
+		if err := rlp.DecodeBytes(bb, &block); err != nil {
+			return nil, errors.Wrap(err, "[GetBlocksByHashesResponse]")
```
[the remark this answers is not in the document]
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:336]: What is your expectation of error message?

File: p2p/stream/protocols/sync/client.go (hunk 3 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:340]: Yes. Nil block is added as a placeholder. The following equation is strictly defined in verification

```
len(req.bns) == len(resp.blockBytes) == len(resp.sigs)
```

File: p2p/stream/protocols/sync/client.go (hunk 4 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/client.go:152]: Yes. nil is a placeholder equivalent to say "I don't have the block message of the given hash/number".

File: p2p/stream/protocols/sync/const.go (hunk 5 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/const.go:10]: Right, it is better to add some comments here. BTW, max size limit is actually calculated based on this number, which is  just an arbitrary number I believe reasonable.

File: p2p/stream/protocols/sync/protocol.go (hunk 6 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/protocol.go:94]: Let me make them constants. Thanks

File: p2p/stream/protocols/sync/protocol.go (hunk 7 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/protocol.go:196]: Nope. This is actually the backoff logic to handle error case.

If Advertise returns an error, the nextWait will return 0. If this is the case, the node will go into an infinite loop which is a black hole for CPU resource. So I added this minimum to retry in 1 min.

For normal non-error cases, nextWait will always be 6 hours from libp2p.

File: p2p/stream/protocols/sync/stream.go (hunk 8 above)
Comment (JackyWYX) [reply p2p/stream/protocols/sync/stream.go:305]: nil is a placeholder to say: "I don't have the block"

**Pull request merged**